### DISCUSSION TITLE
Why use super in crazy generics module

### DISCUSSION THREAD

**Discussion by homosapi3n** [bobocode-projects/java-fundamentals-exercises#158] [Q&A] [answered]
In MaxHolder or StrictProcessor class, the input param has been defined as Comparable<? super T>. Why use super instead of specific type T i.e. Comparable< T > ? 
Can someone provide an example where each option would work but not other?

Adding the code snippet for more clarity:

public static class MaxHolder<T extends Comparable< T >> {  // vs MaxHolder<T extends Comparable<? super T>>
        private T max;

        public MaxHolder(T max) {
            this.max = max;
        }

        /**
         * Puts a new value to the holder. A new value is stored to the max, only if it is greater than current max value.
         *
         * @param val a new value
         */
        public void put(T val) {
            if(val.compareTo(max) > 0)
                max = val;
        }

        public T getMax() {
            return max;
        }
    }

**Comment by aquariusmaster** [comment, marked as answer]
Suppose we have next classes:
```java
class A implements Comparable<A>
class B extends A
```
With `MaxHolder<T extends Comparable< T >>` you will not be able to put `B` in MaxHolder, because `B` does not implement `Comparable<B>`, it implements `Comparable<A>`.
By using `MaxHolder<T extends Comparable<? super T>>` you can do that because MaxHolder will accept `Comparable<A>`.
May be useful: https://stackoverflow.com/a/25783345

**Reply by homosapi3n** [reply]
That makes sense. Thanks for clarifying
